### DISCUSSION TITLE
Redo OCR for single or selected documents

### DISCUSSION THREAD

**Discussion by tpre** [paperless-ngx/paperless-ngx#405] [Feature Requests] [answered]
When you import documents which already contain a text layer, there might be the wish to redo ocr. Currently there is no possibility to redo ocr and update database.

Would be a very nice feature!

**Comment by FrankStrieter** [comment]
@tpre Just to be sure: is this the option you want? https://paperless-ng.readthedocs.io/en/latest/configuration.html 

see PAPERLESS_OCR_MODE -> redo/force

**Reply by sbrunner** [reply]
This is for the consumer, what about already imported document?

**Comment by tpre** [comment]
1) Redo OCR after import for a single document or for multiple documents

Why? Because as default behaviour there is no need to get rid of an existing text layer in general. But if there are documents with bad text-layer, i want to be able to recreate this text-layer.

2) It would also be nice to rebuild database in the way that ocr-text, which is stored in the database can be reread from already stored original files after mass reocr of documents which is done on a more powerful computer, where as paperless-server is to less powerful for mass actions.

**Comment by frrad** [comment]
I would like this option also. For the majority of my documents the default `PAPERLESS_OCR_MODE` works great. However, a few documents already have a (very garbled) text layer which I would like to be able to replace.

**Comment by rantwolf** [comment]
Hi. 
I'm new to paperless-ngx.
Setup was last week in a proxmox/lxc with ubuntu 22.04 (the documentation is very good).
I'm running paperless without docker.

But I made a mistake. In my configuration I forgot to set the option for a full ocr scan.
Now I have over 600 documents in the database which only the first page have ocr.

How do I make a 'ocr-rerun' over all documents without docker?

**Comment by titro** [comment]
So you configured PAPERLESS_OCR_PAGES=1  (Which ist not the default?) Change this to 0 and then:
Use the document_archiver:
https://paperless-ng.readthedocs.io/en/latest/administration.html?highlight=document_archiver#creating-archived-documents

with:
document_archiver --overwrite --document "id"

or for all without defining a document id

**Reply by rantwolf** [reply]
PAPERLESS_OCR_PAGES=1 is default.
Now after importing the first documents I realized that this is not the best option for me.
I changed this param to: PAPERLESS_OCR_PAGES=0

The big question is:
I have a bare metal installation and I do not see any python environment.
If I go to `src/` and call: `python3 manage.py document_archiver --overwrite` I get an error: 

`Traceback (most recent call last):
  File "/opt/paperless/src/manage.py", line 9, in <module>
    from django.core.management import execute_from_command_line
ModuleNotFoundError: No module named 'django'
`

I'm so confused...

**Reply by rantwolf** [reply]
OOHHHH sorry, It's my fault.
I forgot to change the user....
with `sudo -Hu paperless .....` it's working.

Thanks again anyway.

**Comment by GooRoo** [comment]
I installed Paperless-ngx yesterday in docker on my NAS. I already had a bunch of docs from my scanner, most of them were OCR'd with scanner's software. I noticed though that in some of the docs the language was incorrect and the OCR text contains some garbage now. 

The suggested feature would help me to re-run the OCR on selected documents only.

### Addition to the original feature request:
I'd also add that a **user should be able to choose a language** to redo the OCR right in the UI (among the available ones of course).

**Reply by tooomm** [reply]
Another idea:
In combination with https://github.com/paperless-ngx/paperless-ngx/pull/564, it would be helpful to manually rerun OCR on protected PDFs that way after providing the password.

**Comment by mreitschuster** [comment]
I would also need this option.

assume older pdf have been ocr-ed with old versions of ocr software, maybe a different tool, not paperless-ngx. Old OCR might not be great especially on old bad scans. 

So for the majority of my documents i do not want the consumer to redo them, but there might be a few single documents on which I want to reOCR after seeing that the content is not what is expected. 

I would petition for a button in the document viewer's "content" tab, as this is the place where you manually overwrite the OCRed text already.

**Comment by flesser** [comment]
I do have the same problem: Several of my documents already have a very crappy text layer from earlier OCR attempts.
I tried setting `PAPERLESS_OCR_MODE=redo` and `manage.py document_archiver --overwrite`, but then the actual text content of my electronic PDFs got lost and was replaced by an OCRed version of the text in the embedded images only (e.g. company logos).

**Additional feature request for redoing OCR:** being able to specify a different language for selected documents.
Use case: all my documents are in German, but there is that one travel invoice in French...

**Comment by sarcoplasm68** [comment]
Crappy earlier OCR seems to be a common problem. OCRmyPDF provides a _OcrQualityDictionary_ `class` (via `import quality from ocrmypdf`), that may be used together with language specific wordlists, e.g. an `ASPELL` dictionary. _measure_words_matched_ compares words in OCRed text against the dictionary and returns a good guess wether the current OCR is garbage. 
Could be used for conditional redo of OCR.

**Comment by tooomm** [comment, marked as answer]
This is now implemented in [v1.8.0](https://github.com/paperless-ngx/paperless-ngx/releases/tag/v1.8.0) (#1139, #1277), thanks to @stumpylog and @shamoon.

**Comment by timaschew** [comment]
For me the redo OCR is not enough because I can't change OCR Mode or language whicih is somtimes very useful.

I've created a new discussion https://github.com/paperless-ngx/paperless-ngx/discussions/1466
